### ISSUE TITLE
ModelState cannot be added to TempData (cannot be serialized)

### ISSUE THREAD

**Issue by Bartmax** [dotnet/aspnetcore#4816]
I found this when dealing with `ModelState` and transfering using `TempData` to follow `PRG` pattern.

``` c#
TempData.Add("ModelStateTransfer", ModelState);


InvalidOperationException: 
The 'Microsoft.AspNet.Mvc.ViewFeatures.SessionStateTempDataProvider' cannot serialize an object
of type 'Microsoft.AspNet.Mvc.ModelBinding.ModelStateDictionary' to session state.
```

Shouldn't be `ModelState` serializable?

I was pointed out to this tests:
https://github.com/aspnet/Mvc/blob/25eb50120eceb62fd24ab5404210428fcdf0c400/test/Microsoft.AspNetCore.Mvc.Core.Test/SerializableErrorTests.cs

so I tried using `SerializableError` with no success

``` C#
TempData.Add("ModelStateTransfer", new SerializableError(ModelState));

InvalidOperationException: 
The 'Microsoft.AspNet.Mvc.ViewFeatures.SessionStateTempDataProvider' cannot serialize an object
of type 'Microsoft.AspNet.Mvc.SerializableError' to session state.
```

Help? Ideas?

**Comment by bsimser** [comment]
I just hit this issues trying to use what I felt was a pretty good implementation of PRG with Toastr. The original article can be found [here](http://toplinestrategies.com/blogs/net/standardizing-aspnet-mvc-feedback-toastr). It doesn't work with core because a) there's no ControllerExtensions class that will let you do a generation RedirectToAction and b) if you manually do the PRG pattern and stuff the toast message (using the alert class) directly into TempData you get this error. The [workaround on StackOverflow](https://stackoverflow.com/questions/34638823/store-complex-object-in-tempdata) works but is that the answer or is there going to be something better built-in?

**Comment by Neme12** [comment]
It would be nice if TempDataSerializer at least supported tuples.

**Comment by Perustaja** [comment]
Long time since but huge plus on this.
@bsimser I saw this code all over too. Personally I just find this incredibly confusing and cannot imagine stumbling upon this in someone else's controllers and views. It encourages developers to use TempData in a very odd way and anyone trying to use it normally down the line will have to follow the breadcrumb trail and figure out what's going on. We're left with either bypassing TempData's expected usage or bloating the controller/page with properties.

**Comment by Yaevh** [comment]
Andrew Lock provides a workaround for this problem in MVC: https://andrewlock.net/post-redirect-get-using-tempdata-in-asp-net-core/
Unfortunately his method doesn't work in Razor Pages, so I've created a `IPageFilter` for this, maybe it'll be helpful for someone: https://gist.github.com/Yaevh/e87f682a3c3ac35d1504c068c9f5e8ab

I however agree that it should be supported by the framework itself and not require additional work by the developer.

**Comment by sipi41** [comment]
@Yaevh sorry, I'm newbie, would you please explain how to use your IPageFilter? is there any example? I understand we must create a class with what you wrote, then add this to Startup.cs but later? how do implement this? thank you for your patience and help.

**Comment by sipi41** [comment]
I guess I haven't read about filters :-( I have done some validation attributes but not a filter... :-) but I'm learning and very excited about learning, sometimes I feel like I will never stop learning... I'm on the 22 (of 40 chapters) of Professional c#, have almost completed a book of Entity Framework and now I did stop reading Microsoft Blazor... too much to learn, and they are producing new things every day, thanks for your answer @Yaevh 

I think I found a solution... First I was thinking that if action methods are methods in a class, they share fields and/or properties, so by creating a field of type ModelStateDictionary, we could store the value there then retrieve it and vuala!, unfortunately, I did notice that it does save the data there but after calling the form method, the field/property is reset to null again... 

**SOLUTION**: I was thinking... what if we take another approach? something similar that doesn't depend on an instance of the class but belongs to the class, and then I remember we could use a static field, and it did the work!

`
public class AdministrationController : Controller
    {
        public static ModelStateDictionary ModelStateDic { get; set; } = null;

        private UserManager<Usuario> usersManager { get; set; }
        private AppDbContext db { get; set; }

        public AdministrationController(UserManager<Usuario> userMgr, AppDbContext Db)
        {
            this.usersManager = userMgr;
            this.db = Db;
        }

        public IActionResult Main() => View();
        

        public async Task<IActionResult> ManageUsers() {            
            
            ViewBag.usuarios = await db.Users.Include(e => e.UserRoles).ToListAsync();

            ViewBag.roles = await db.Roles.Select(s => new SelectListItem { Text = s.Name, Value = s.Id }).ToListAsync();

            if (AdministrationController.ModelStateDic != null)
            {
                ModelState.Merge(AdministrationController.ModelStateDic);                
            }

            return View(model: new CreateNewUser());
        
        }

        [HttpPost, ValidateAntiForgeryToken]
        public async Task<IActionResult> CreateNewUser(CreateNewUser NuevoUsuario) {


            AdministrationController.ModelStateDic = null;

            ModelState.AddModelError("ErrorAddingUser", "ERROR: error sample msg.");

           AdministrationController.ModelStateDic = ModelState;

            return RedirectToAction("ManageUsers");
        
        }

    }`

In other words, instead of saving the data in Temp, we use a local static field to save the data.

**Comment by leniency** [comment]
> In other words, instead of saving the data in Temp, we use a local static field to save the data.

@sipi41 - the problem with this though is that static variables is that they remain in memory past the current request - they retain value for the lifetime of the **application**. It may work in test with a single user, but with multiple users, this is very dangerous. _Every_ user will see the same value for that static variable. Not to mention potential concurrency issues.

**Comment by sakopov** [comment]
Posting for anyone looking for a proper solution to this.

```csharp
  public class ModelStateEntryDto
  {
      public string Key { get; set; }

      public string AttemptedValue { get; set; }

      public object RawValue { get; set; }

      public ICollection<string> ErrorMessages { get; set; } = new List<string>();
  }
```

Create custom action filter attribute to transfer model state to temp data. This should be added to any action that modifies `ModelState` and redirects.

```csharp
    public class TransferModelStateAttribute : ActionFilterAttribute
    {
        private const string ModelStateTempDataKey = "ModelState";

        public override void OnActionExecuted(ActionExecutedContext filterContext)
        {
            var controller = filterContext.Controller as Controller;

            if (controller != null)
            {
                var modelStateEntries = controller.ViewData.ModelState
                    .Select(modelStateEntry => new ModelStateEntryDto
                    {
                        Key = modelStateEntry.Key,
                        AttemptedValue = modelStateEntry.Value.AttemptedValue,
                        RawValue = modelStateEntry.Value.RawValue,
                        ErrorMessages = modelStateEntry.Value.Errors.Select(err => err.ErrorMessage).ToList()
                    });

                controller.TempData[ModelStateTempDataKey] = JsonSerializer.Serialize(modelStateEntries);
            }

            base.OnActionExecuted(filterContext);
        }
    }
```

Create another action filter attribute to restore model state from temp data. This should be added to any action that needs to display data in `ModelState` (ie validation summary).

```csharp
    public class RestoreModelStateAttribute : ActionFilterAttribute
    {
        private const string ModelStateTempDataKey = "ModelState";

        public override void OnActionExecuting(ActionExecutingContext filterContext)
        {
            var modelState = new ModelStateDictionary();

            var controller = filterContext.Controller as Controller;

            if (controller?.TempData != null && controller.TempData.TryGetValue(ModelStateTempDataKey, out object serializedModelState))
            {
                var modelStateEntries = JsonSerializer.Deserialize<List<ModelStateEntryDto>>((string)serializedModelState);

                foreach (var error in modelStateEntries)
                {
                    modelState.SetModelValue(error.Key, error.RawValue, error.AttemptedValue);

                    foreach (var errorMessage in error.ErrorMessages)
                    {
                        modelState.AddModelError(error.Key, errorMessage);
                    }
                }

                controller.ViewData.ModelState.Merge(modelState);
            }

            base.OnActionExecuting(filterContext);
        }
    }
```

**Comment by Atulin** [comment]
Just stumbled upon the same limitation in my Razor project. @Yaevh's solution seems to be working perfectly fine, but it would be really nice to have built-in support for PRG.

**Comment by mkArtakMSFT** [comment]
Thanks for contacting us.
We couldn't prioritize this investigation given the number of customers are affected. Hence closing this issue.
You can learn more about our triage process and how we handle issues by reading our [Triage Process](https://github.com/dotnet/aspnetcore/blob/master/docs/TriageProcess.md) writeup.
